addons/udes_stock/models/stock_picking.py

```diff
@@ -465,37 +465,59 @@ def action_done(self):
         Ensure we don't incorrectly validate pending pickings.
         Check if picking batch is now complete
         """
-        self.assert_not_pending()
-        mls = self.mapped("move_line_ids")
-        # Prevent recomputing the batch stat
-        batches = mls.mapped("picking_id.batch_id")
-        next_pickings = self.mapped("u_next_picking_ids")
-
-        res = super(StockPicking, self.with_context(lock_batch_state=True)).action_done()
-
-        # just in case move lines change on action done, for instance cancelling
-        # a picking
-        mls = mls.exists()
-        picks = mls.mapped("picking_id")
-        # batches of the following stage should also be recomputed
-        all_picks = picks | picks.mapped("u_next_picking_ids")
-        all_picks.with_context(lock_batch_state=False)._trigger_batch_state_recompute()
-
-        self.assert_not_lot_restricted()
-
-        extra_context = {}
-        if hasattr(self, "get_extra_printing_context"):
-            extra_context = picks.get_extra_printing_context()
-
-        # Trigger print strategies for pickings done
-        self.env.ref("udes_stock.picking_done").with_context(
-            active_model=picks._name,
-            active_ids=picks.ids,
-            action_filter="picking.action_done",
-            **extra_context
-        ).run()
+        res = False
+        # Collect action_done stats for monitor
+        with self.statistics() as stats:
+            self.assert_not_pending()
+            mls = self.mapped("move_line_ids")
+            # Prevent recomputing the batch stat
+            batches = mls.mapped("picking_id.batch_id")
+            next_pickings = self.mapped("u_next_picking_ids")
+
+            res = super(StockPicking, self.with_context(lock_batch_state=True)).action_done()
+
+            # just in case move lines change on action done, for instance cancelling
+            # a picking
+            mls = mls.exists()
+            picks = mls.mapped("picking_id")
+            # batches of the following stage should also be recomputed
+            all_picks = picks | picks.mapped("u_next_picking_ids")
+            all_picks.with_context(lock_batch_state=False)._trigger_batch_state_recompute()
+
+            self.assert_not_lot_restricted()
+
+            extra_context = {}
+            if hasattr(self, "get_extra_printing_context"):
+                extra_context = picks.get_extra_printing_context()
+
+            # Trigger print strategies for pickings done
+            self.env.ref("udes_stock.picking_done").with_context(
+                active_model=picks._name,
+                active_ids=picks.ids,
+                action_filter="picking.action_done",
+                **extra_context
+            ).run()
+
+        # Write the log
+        # Ensure when multiple picking types are involved that the names are combined
+        picking_type_names = (
+            self.picking_type_id.name
+            if len(self) == 1
```

Comment:
Why do we need this if? Join shouldn't add "," if there is only one element in the list.

Reply:
Addressed